### PR TITLE
feat: make sure service level updates username

### PR DESCRIPTION
**Pull request by remisultan**
## :id: Reference related issue. 

https://gravitee.atlassian.net/browse/AM-445

## :pencil2: A description of the changes proposed in the pull request

This PR ensures that the username is update where necessary if possible

## :memo: Test Scenarios

`curl -vvv -X POST http://localhost:8093/management/auth/token -H 'authorization: Basic base64(username:password)'`

Will get you an access token

```curl -vvv -X PATCH -H "content-type: application/json" -H "Authorization: Bearer <access-token>" http://localhost:8093/management/organizations/DEFAULT/environments/DEFAULT/domains/<domainId>/users/<userId>/username -d '{ "username"  : "<newUsername>" }' ```

- Test if the username respects the username pattern -- should fail
- Test if the username already exists -- should fail
- Finally update a user with its actual username -- should succeed
- Try updating the username of a non UserProvider IDP (Everything but Mongo/Jdbc/Http) -- should fail

### CODE REVIEW

File: gravitee-am-identityprovider/gravitee-am-identityprovider-jdbc/src/main/java/io/gravitee/am/identityprovider/jdbc/user/JdbcUserProvider.java [hunk 1]
```diff
@@ -343,6 +343,32 @@ public Single<User> update(String id, User updateUser) {
                 });
     }
 
+    @Override
+    public Completable updateUsername(String id, String username) {
+        if (Strings.isNullOrEmpty(username)) {
+            return Completable.error(new IllegalArgumentException("Username required for UserProvider.updatePassword"));
```
Comment (mariano-fernandez) [inline gravitee-am-identityprovider/gravitee-am-identityprovider-jdbc/src/main/java/io/gravitee/am/identityprovider/jdbc/user/JdbcUserProvider.java]: `UserProvider.updatePassword`? I'm guessing a small copy/paste oversight 😜

File: gravitee-am-identityprovider/gravitee-am-identityprovider-api/src/main/java/io/gravitee/am/identityprovider/api/UserProvider.java [hunk 2]
```diff
@@ -37,6 +37,8 @@ default Maybe<User> findByEmail(String email) {
 
     Single<User> update(String id, User updateUser);
 
+    Completable updateUsername(String oldUsername, String username);
```
Comment (mariano-fernandez) [inline gravitee-am-identityprovider/gravitee-am-identityprovider-api/src/main/java/io/gravitee/am/identityprovider/api/UserProvider.java]: The first parameter is called `oldUsername` in the interface and `id` in all implementations
I think they should be in line

Comment (remisultan) [reply gravitee-am-identityprovider/gravitee-am-identityprovider-api/src/main/java/io/gravitee/am/identityprovider/api/UserProvider.java]: This is an old revision. 
The last commit 

```
 default Single<User> updateUsername(User user, String username) {
        if (username == null || username.isEmpty()) {
            return Single.error(new IllegalArgumentException("Username required for UserProvider.updateUsername"));
        }
        ((DefaultUser) user).setUsername(username);
        return this.update(user.getId(), user);
    }
```

File: gravitee-am-identityprovider/gravitee-am-identityprovider-jdbc/src/main/java/io/gravitee/am/identityprovider/jdbc/user/JdbcUserProvider.java (hunk 1 above)
Comment (remisultan) [reply gravitee-am-identityprovider/gravitee-am-identityprovider-jdbc/src/main/java/io/gravitee/am/identityprovider/jdbc/user/JdbcUserProvider.java]: Definitely ! 😆

**Review by aurelien-pacaud — APPROVED** [review]

**Pull request merged**